### DISCUSSION TITLE
Multiple ev3 Ultrasonic Sensors does a call to distance(silent=True) block

### DISCUSSION THREAD

**Discussion by gregorianrants** [pybricks/support#1644] [Q&A]
I would like to build an obstacle avoiding robot with multiple ev3 ultrasonic sensors probably 3 or 4.
however i am struggling to find enough information about how the ev3 ultrasonic sensors work to figure out how i would do this.
I dont currently have the sensors so cant run any experiments. i am trying to figure out if my idea is feasable before i buy them.
Any help would be greatly appreciated.

I know that i need to ping sensors at different times so they dont interfere. acording to the docs i would do distance(silent=True).
But i also know that an US sensor takes a while to ping get a an echo back and figure out a distance.  i will need to do other things while it awaits a reading.
if i call distance does it:
-  block till a reading comes back, thus preventing my code from doing other things?
- or
- return null if there is no result yet, thus allowing other things to be done in polling loop or some other cooperative multitaksing technique?

also regarding setting silent to True the docs says  "If you do this too frequently, the sensor can freeze. If this happens, unplug it and plug it back in". how frequently is too frequently, would continously reading one sensor giving it a break while reading another then reading the first again be too much?

Also i am wondering does all the distance calculation happen on the sensor and the python code just recieves the distance or is the UltrasonicSensor class doing pinging and listening for echos, under the hood much like you would with a US sensor from rapspberry pi or arduino.

**Comment by dlech** [comment]
Ultrasonic sensors aren't great for obstacle avoidance (unless somehow all obstacles have large flat surfaces exactly parallel to the face of the ultrasonic sensors).  For the cost of 4 EV3 ultrasonic sensors, you could probably get lidar that would work much better.

Here is an old project for inspiration: https://www.ev3dev.org/docs/tutorials/using-xv11-lidar/

**Reply by gregorianrants** [reply]
Thanks that looks interesting. 

I did discover the hard way that ultrasonic sensors are not as easy as a lot of the projects on you tube would suggest.  However after a lot of effort i did figure out how to do robust obstacle avoidence with 5 US sensors attached to a raspberry pi.

you can see in in action here [buildhat_node_bot](https://github.com/gregorianrants/buildhat_node_bot)
it uses the build hat to power lego motors, the build hat turned out to be pretty buggy and i had to reimplement the pid speed control in software on the pi.  would be great if pybricks had a library for the build hat.

my understanding is that lidar only measures distances on a plane. the thing i like about the US sensors is that they can detect things like floating shelves etc.
i was keen to try and do the same thing with nothing but lego kit.  
but would need to be able to 

1. ping sensors at different times without them feezinng
2. run other code while waiting for a US sensor to return a value. when using silent mode.

re price: i just looked and yes the ev3 sensors too pricey to buy multiples. i have opted to go for nxt ones.
I have managed to pick up 3 nxt US sensors for £12 on bricklink so will mess about with them and report back if anything comes of it.

**Reply by dlech** [reply]
I'm impressed!

Good idea with the NXT sensors. They will probably work better anyway. 

It's been too long, I don't remember exactly how the "Single Measurement" mode work, but hopefully that is what you are looking for.

**Reply by gregorianrants** [reply]
thanks for your feedback, i will update this thread when i have investigated and found out more.

**Reply by gregorianrants** [reply]
I have came back to this again and have done some digging

the way that this function works is very dependant on how the ultrasonic sensor works which is a bit awkward

in ev3dev terms you write to the mode US-SI-CM this triggers a ping and when the ping has finished it updates the value held in the sensor and the value doesnt change till it does another ping.

when you use the function distance(silent=True), in pybricks, it will make a ping then return whichever value is held on the sensor without any blocking. since we should allways have  delay between triggering pings the value that you will get will be the value that was the result of the previous ping. i.e. the distance you get is not from the ping that was made in the function call that returned it but the distance that was measured on a ping that was made on a prior call of distance(silent=True).  The first time you use distance(silent=True) you will get 0 because there was no previous ping.

it may seem that there should be a function/property that gives us the value seperately from triggering a ping. but i can see why this is not implemented.  it would be difficult to tell when we have a new value. my first thought was to just check if the value has changed but sometimes consecutive pings will give the same distance. so there is no simple way of doing this.

ev3dev does not have a way of checking the value in the api. but you can dig into the private methods and properties and do this. it could also be done via sysfs. one way of checking if we have new value is to check if it has changed or if a timeout has passed. the time out would have to be based on the previous distance and the speed of sound. i.e. if the last distance was 50cm and enought time has passed for a sound wave to travel 100cm (50cm out and 50 back). then we probably have a new reading. i may mess about with trying to do this via sysfs and create a class for some more advanced usage.  however this may not be necessary and if i can get an obstacle avoiding robot working with 4 us sensors without over complicating things that will probably do.

another thing i have discovered is how frequently you can ping. ev3dev says that the sensor can lock up if you ping to much and recomends 250ms. this is quite long for robotics and surprsising. the max range is 255cm this would take 14ms for the ping out and back. i ran the sensor for quite a while with a delay of 50ms between pings with no problems. i havent experimented with a shorter delay yet.

**Reply by laurensvalk** [reply]
Hey @gregorianrants. We've got a brand new version of Pybricks for EV3 coming up. You can try a very early version [here](https://labs.pybricks.com/). Would you like to try and test how we can improve the ping/listen feature? This time, we have direct control over the sensor so things should be very fast (only constrained by the sensor itself).

You can access modes like this.

```python
from pybricks.ev3devices import UltrasonicSensor
from pybricks.parameters import Port
from pybricks.tools import wait
from pybricks.iodevices import LUMPDevice

device = LUMPDevice(Port.S4)
print("INFO:", device.info())

while True:
    data = device.read(0)
    print(data)
    wait(100)
```

```
INFO: {'id': 30, 'modes': (('US-DIST-CM', 1, 1), ('US-DIST-IN', 1, 1), ('US-LISTEN', 1, 0), ('US-SI-CM', 1, 1), ('US-SI-IN', 1, 1), ('US-DC-CM', 1, 1), ('US-DC-IN', 1, 1))}
(353,)
(326,)
(326,)
(59,)
(30,)
(50,)
(109,)
(251,)
(302,)
```

What are the things we need to write? `LUMPDevice` has a write method, but I don't think this sensor self-reports to be writeable.

Or is switching to that mode (by doing a read) enough to trigger the ping? And then how do we read the value?

If you can create something reliable with the example above, we could include the same functionality in the standard `UltrasonicSensor` class.

**Reply by BertLindeman** [reply]
code example needs a '#' on the comment line (line 8) and the line is incomplete 😄

**Reply by gregorianrants** [reply]
Hey Laurens,  I would find it interesting to have a go at this and will see what I can do.

I probably worded something wrongly above 
I said when "you write to the mode US-SI-CM this triggers a ping"
I should probably have said when you select a mode.

I may not be correct to say you can trigger and ping and read a value separately.  I was looking at how ev3dev works and certainly there are options to set the mode of the sensor and get the last measured value separately from each other, in python or sysfys. however I may be getting misdirected by the abstraction that ev3dev puts over the sensor hardware.

you can see this in pybricks ev3 v1
https://github.com/pybricks/pybricks-api/blob/master/src/pybricks/iodevices.py

where 

class Ev3devSensor():
 
has methods

```

    def _mode(self, mode_new):
        """Set the sensor mode. Not available to user."""
        if mode_new != self.mode_now:
            write_str(self.mode_file, mode_new)
            self.mode_now = mode_new

    def _value(self, num):
        """Return value in sensor/valueX. Not available to user."""
        return read_int(self.value_files[num])

```
I am starting by looking at the uart protocol

I had seen that on ev3dev website it said
that the us sensor does not support commands

my hope was that it meant that it didn’t have any sensor specific commands such as with the gyro which has
BEGIN-COMP-CAL for begin compass calibration

but that it would have some commands that were common to all sensors such as from the uart protocol 
CMD_SELECT which selects a mode

more info here
https://github.com/pybricks/technical-info/blob/master/uart-protocol.md
 
I am also trying to figure out if uart sensors continously send data
if so we must have to send some sort of command to change the mode

where as if we request data then I guess when we request data from a specific mode that could change the mode

one theory I have is that the command you use  above
read(0)
checks we are on the right mode if not changes it then 
then picks up the next reading which is broadcast.


but that maybe the sensor actually has a way of setting the mode separately.

I am trying to dig into the pybricks code but I am just a hobbyist and havent really dealt with c code of that scale before.
I mostly do python  and my experience of c is just arduino.

I will keep plugging away and see what I can do.

**Reply by laurensvalk** [reply]
Right, just reading from a specific mode will send the command to switch the mode. The above example switches to and then reads from mode 0.

So perhaps you could try experimenting with the other listed modes to see what exactly they produce?

**Reply by gregorianrants** [reply]
ill have a go at it

**Reply by gregorianrants** [reply]
I have got some tentatively encouraging preliminary results regarding being able to ping and read seperately.

first i ran your example and it behaved as expected, i varied the distance of an object in front of the sensor and the values read varied in tandem.

then i experimented with US-SI-CM and it did not behave as expected.

```
device = LUMPDevice(Port.S1)
print("INFO:", device.info())


while True:
    dist = device.read(3)
    print(dist)
    wait(500)
```

this gives the same measurement for every call to read. the physical distance to object in front of sensor is varied but the reading doesnt change.

```
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
(446,)
```

most of the info i can find about how the ultrasonic sensor works says that setting the mode to US-SI-CM triggers a  single ping and that you set the mode on the sensor to US-SI-CM again (although thats the mode it is already in).  

perhaps the pybricks code is checking what mode the sensor is in by checking some state variable in software and then neglecting to set it becuase its already set.

another possibility is that the sources i have been looking at are not correct and that setting the mode to US-SI-CM actually stops the sensor pinging and you rely on a ping from setting briefly to conitinous mode to cause the ping.

an old post from dlech in 2014 suggests this:
he used a logic analyzer to to analyze the what the ev3-g does for single shot and said "It turns out that the LEGO VM the single ultrasonic sensor block actually sets the mode to US-DIST-CM and then sets the mode to US-SI-CM twice! "

`https://github.com/ev3dev/ev3dev/issues/154#issuecomment-56910497`

either way i decided to see what i could get out of it. perhaps i have stumbled on how the sensor works on the other hand i may just be exploiting a bug.

i tried the experiment bellow

```
device = LUMPDevice(Port.S1)
#change mode immediately to US-SI-CM (single shot mode) to stop the sensor pinging
device.read(3)
print("INFO:", device.info())
while True:
    # while len(ev3.buttons.pressed()) ==0:
    #     wait(100)
    print('new')
    #change mode to continous to trigger a ping
    dist = device.read(0)
    #allow enough time for a ping out and back, bit of a guess that it wont ping again within this time
    wait(15)
    #change mode to single shot and get the reading from the ping
    dist = device.read(3)
    print(str(dist))
    wait(500)
    #read from single shot again value doesnt change
    dist = device.read(3)
    print(str(dist))
    wait(500)
    #read from single shot again value doesnt change
    dist = device.read(3)
    print(str(dist))
    wait(500)
    #read from single shot again value doesnt change
    dist = device.read(3)
    print(str(dist))
    wait(500)

```

and it behaved in line with my theory

```
INFO: {'id': 30, 'modes': (('US-DIST-CM', 1, 1), ('US-DIST-IN', 1, 1), ('US-LISTEN', 1, 0), ('US-SI-CM', 1, 1), ('US-SI-IN', 1, 1), ('US-DC-CM', 1, 1), ('US-DC-IN', 1, 1))}
new
(1555,)
(1555,)
(1555,)
(1555,)
new
(63,)
(63,)
(63,)
(63,)
new
(51,)
(51,)
(51,)
(51,)
new
(74,)
(74,)
(74,)
(74,)
new
(50,)
(50,)
(50,)
(50,)
new
(2374,)
(2374,)
(2374,)
(2374,)
new
(1555,)
The program was stopped (SystemExit).
```
 
of course there could still be extra pings in there although the reading doesnt change. so i created a class to encaspsulate things and ran a trial with 2 sensors pinging alternately. there was no evidence of interference.  the sensors are right next to each other and facing the same direction.


```

class SingleShot:
    def __init__(self,port):
        self.device = LUMPDevice(port)
        self.device.read(3)

    def read(self):
        self.device.read(0)
        wait(15)
        dist = self.device.read(3)
        return dist

us1 = SingleShot(Port.S1)
us2 = SingleShot(Port.S2)


while True:
    dist = us1.read()
    print('s1',dist)
    wait(500)
    dist = us2.read()
    print('s2',dist)
    wait(1000)
```

```
s1 (198,)
s2 (191,)
s1 (190,)
s2 (190,)
s1 (362,)
s2 (410,)
s1 (391,)
s2 (392,)
s1 (384,)
s2 (389,)
s1 (237,)
s2 (179,)
s1 (175,)
s2 (174,)
s1 (175,)
s2 (170,)
s1 (462,)
s2 (464,)
s1 (460,)
s2 (456,)
s1 (446,)
s2 (443,)
The program was stopped (SystemExit).
```


i am holding a board in front of the sensors sometimes at a fixed distance and sometimes moving it. you will notice that the s1 and s2 are mostly the same though there will be a discrepancy sometimes as i move the board.

if both sensors were pinging continously i think there would be highly spurious readings due to interference.
if of interest i can give a more controlled demonstration with a video or something so the phyiscal set up can be seen at the same time as the readings.
